### ISSUE TITLE
Feature request: Mark comment as untouched

### ISSUE THREAD

**Issue by dalgard** [kangax/html-minifier#70]
It would be great to have the ability to strip comments but to leave out specially marked banner comments (for instance). I don't suppose this can be achieved already?

**Comment by kangax** [comment]
That's a reasonable idea. Marking as possible feature.

**Comment by mhulse** [comment]
I'd love to see this option too. Maybe something like in CSS comments:

``` css
/*!
 * This is a comment that's untouched by the minification process.
 */
```

Single line:

``` css
 /*! This is a comment that's untouched by the minification process. */
```

For html comments, maybe something like:

``` html
<!--! This is a comment that's untouched by the minification process. -->
```

... or something like that. :+1:

My goal would be to include a banner at the top of my `<head>` ... I'd like to minify everything else and keep the banner. :smile:

**Comment by kangax** [comment]
@mhulse What do you mean by a banner?

**Comment by mhulse** [comment]
@kangax Good question!

I think it would be really awesome to put "banner" comments into my compressed HTML.

In my particular case, I'm wanting to go **l33t** using something like [grunt-asciify](https://github.com/olizilla/grunt-asciify) and do something like:

``` html
<!DOCTYPE html>
<!--!
  ______            _                 _                   
 |  ____|          | |               | |                  
 | |__ ___   ___   | |__   __ _ ____ | |__   __ _ _ __    
 |  __/ _ \ / _ \  | '_ \ / _` |_  / | '_ \ / _` | '__|   
 | | | (_) | (_) | | |_) | (_| |/ /  | |_) | (_| | |      
 |_|  \___/ \___/ _|_.__/ \__,_/___| |_.__/ \__,_|_|      
      /_ | / _ \ / _ \                                    
 __   _| || | | | | | |                                   
 \ \ / / || | | | | | |                                   
  \ V /| || |_| | |_| |   __ ___      ___   ___  __ ____  
   \_/\|_(_)___(_)___/   /_ |__ \    |__ \ / _ \/_ |___ \ 
    /  \  _   _  __ _     | |  ) |      ) | | | || | __) |
   / /\ \| | | |/ _` |    | | / /      / /| | | || ||__ < 
  / ____ \ |_| | (_| |_   | |/ /_ _   / /_| |_| || |___) |
 /_/    \_\__,_|\__, (_)  |_|____( ) |____|\___/ |_|____/ 
                 __/ |           |/                       
                |___/                                     
-->
<html><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1"> .... </body></html>
```

I'm sure there are better reasons to have this feature, the above is just my personal use case.

A more real world example might be:

``` html
<!DOCTYPE html><!--! Foo bar baz. vX.X.X | Apache License, Version 2.0 | https://github.com/billyjohnson/foo --><html><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1"> .... </body></html>
```

I suppose one could use [`grunt-contrib-concat`](https://github.com/gruntjs/grunt-contrib-concat) to append a banner to the very top of the HTML file (above the doctype), but I remember that older IEs get thrown into quirks mode if there's anything above a DTD. Here's a possible authoritative note about quirksmode behavior:

> [Make sure you put the DOCTYPE right at the beginning of your HTML document. Anything before the DOCTYPE, like a comment or an XML declaration will trigger quirks mode in Internet Explorer 9 and older.](https://developer.mozilla.org/en-US/docs/Quirks_Mode_and_Standards_Mode)
> 
> In HTML5, the only purpose of the DOCTYPE is to activate full standards mode. Older versions of the HTML standard gave additional meaning to the DOCTYPE, but no browser has ever used the DOCTYPE for anything other than switching between quirks mode and standards mode.

... Interesting stuff.

So, that's why the Grunt concat task might not be the best option as one would need to insert the comment after the DTD if they wanted to avoid quirksmode in older IEs (that is, if you consider IE9 old, hehe).

Again, I'm sure other folks have other (better?) reasons to preserve comments (and whitespace within those comments).

**Comment by mhulse** [comment]
I forgot to mention ... Just to clarify, in other build tools, the "`!`" point is the flag used to tell the uglifier/minifier to not strip said comment. Like:

``` css
/*!
 * Description.
 *
 * @blah DocBlock style comments.
 */
```

... or:

``` js
/*! Foo bar baz. vX.X.X | Apache License, Version 2.0 | https://github.com/billyjohnson/foo */
```

... that's why I recommended doing something like:

``` html
<!--! ... -->
```

Though, maybe there's a better way to flag HTML comments?

Hmm, I wonder if you could just do a replacement on the comment for the output version? Like, for example, this:

``` html
<!--! ... (comment with optional whitespace/newlines) -->
```

... would get converted to this:

``` html
<!-- ... (preserved whitespace/newlines) ... -->
```

... in the final minified version?

**Comment by kangax** [comment]
Ok, that's what I thought you meant by "banner" :)

So yeah, the main use case for this is license; and yes, following `/*!` convention makes sense. I like it.

I don't have time for this (and really don't know when I will have it) so if anyone is willing contribute, I'd happily pull this in.

**Comment by mhulse** [comment]
> I don't have time for this (and really don't know when I will have it) so if anyone is willing contribute, I'd happily pull this in.

I'd be willing to do a PR. I'd need to wrap my head around your code in order to get a feel for how things work, but once that happens, I'd give it my best shot.

Off the top of your head, do you have any tips on which files I'd need to modify and what methods I should focus on? I only ask, as it might speed up the time it takes to learn how your code works. :+1: 

Also, does it make sense to look for `<!--! ... -->` and then converting it to `<!-- ... -->` for the final compressed version; or, is there a better way of flagging HTML comments that we haven't thought of in this thread?

**Comment by zcorpan** [comment]
Removing the ! means that minifying again would drop the comment, which seems like a bad idea.

**Comment by zcorpan** [comment]
It seems to me this would be implemented by inserting an else if here https://github.com/kangax/html-minifier/blob/gh-pages/src/htmlminifier.js#L402

**Comment by mhulse** [comment]
@zcorpan 

> Removing the ! means that minifying again would drop the comment, which seems like a bad idea.

Good call. Appreciate the clarification. :+1: 

> It seems to me this would be implemented by inserting an else if here https://github.com/kangax/html-minifier/blob/gh-pages/src/htmlminifier.js#L402

Thanks for tip! I'll try to send a PR this weekend! :smile:

**Comment by mhulse** [comment]
Pull request has been submitted.

I've added a few commit comments:

https://github.com/mhulse/html-minifier/commit/b1580b2763fd5b72e42e3212257a1201d7239c81

### Questions:
1. Is there anything I need to edit in the `master` branch?
2. I wasn't sure how to run the `v8` test. I don't have v8 installed (though, I do have `node.js` installed). Can I homebrew install `v8`? I wasn't able to get the v8 JS test script to run using node command line. Anyway, Let me know if you want me to get that setup and tested before you pull (I should say, if you decide to pull) this PR into your `gh-pages` branch.
3. Is there anything I'm missing in terms of files to updated? I just manually copied JS from `htmlminifier.js` into `all.js`. I assume that's the normal workflow? :smile:

Anyway, the patch was pretty simple. Though, I have to admit, I have not done a ton of testing.

The test suite page passed all tests.

Thoughts?

**Comment by kangax** [comment]
@mhulse thanks!

> Is there anything I need to edit in the master branch?
> Nope.
> 
> I wasn't sure how to run the v8 test. I don't have v8 installed (though, I do have node.js installed). Can I homebrew install v8? I wasn't able to get the v8 JS test script to run using node command line. Anyway, Let me know if you want me to get that setup and tested before you pull (I should say, if you decide to pull) this PR into your gh-pages branch.

Wait, why do you need v8? All the tests are running through node (just `npm test` after you built distribution)

> Is there anything I'm missing in terms of files to updated? I just manually copied JS from htmlminifier.js into all.js. I assume that's the normal workflow? :smile:

all.js is just a concat of 3 files (parser, minifier, lint). I added instructions to README, for future!

**Comment by mhulse** [comment]
> Wait, why do you need v8? All the tests are running through node (just `npm test` after you built distribution)

Ah! It's funny, I tried several ways of testing, but not `npm test`.

Just tested:

``` bash
$ npm install
# ...
$ npm test

> html-minifier@0.5.3 test /Volumes/.../html-minifier
> jshint src/htmlminifier.js src/htmllint.js && cat src/htmlparser.js src/htmlminifier.js src/htmllint.js > dist/all.js && node test.js


Testing /Volumes/storage/Users/mhulse/github/html-minifier/src/htmlminifier.js ... done
Summary:
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┓
┃ File                                               ┃ Failed   ┃ Passed   ┃ Total    ┃ Runtime  ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━┫
┃ ...mhulse/github/html-minifier/src/htmlminifier.js ┃ 0        ┃ 334      ┃ 334      ┃ 77       ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━┛

Global summary:
┏━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Files      ┃ Tests      ┃ Assertions ┃ Failed     ┃ Passed     ┃ Runtime    ┃
┣━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━━┫
┃ 1          ┃ 42         ┃ 334        ┃ 0          ┃ 334        ┃ 77         ┃
┗━━━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━━━┛
```

Zero failures. :+1: 

> `all.js` is just a concat of 3 files (parser, minifier, lint). I added instructions to `README`, for future!

Awesome! Great tip on the file concat too. I just manually copied/pasted, but `> cat src/htmlparser.js src/htmlminifier.js src/htmllint.js > dist/all.js` is so much easier (one of those "why didn't I think of that!" moments). :laughing: 

Let me know if there's anything else I can do to help test.

**FYI:** If/when you pull this request in, I was planning on trying to patch #77. :octocat: 

Thanks @kangax!

**Comment by kangax** [comment]
@mhulse Cool. So I'll close pull request once you add unit tests for this new behavior, as I [mentioned here](https://github.com/kangax/html-minifier/pull/79)

**Comment by mhulse** [comment]
> Cool. So I'll close pull request once you add unit tests for this new behavior, as I mentioned here

Ack, sorry. Missed that. Doing now. :+1:

**Comment by mhulse** [comment]
FYI, this issue has been resolved in 05969fe200f1898985970d4a41db9c46891a41cb.

**Comment by kangax** [comment]
Yep, this is now fixed. Thanks @mhulse!
